File: elearningit/lib/models/call.dart
// models/call.dart
import 'package:json_annotation/json_annotation.dart';
import 'user.dart';

part 'call.g.dart';

@JsonSerializable()
class Call {
  @JsonKey(name: '_id')
  final String id;
  final User? caller;
  final User? callee;
  final String type; // 'voice' or 'video'
  final String
  status; // 'initiated', 'ringing', 'accepted', 'rejected', 'ended', 'missed', 'busy'
  final DateTime? startedAt;
  final DateTime? endedAt;
  final int duration; // in seconds
  final bool isScreenSharing;
  final DateTime? screenShareStartedAt;
  final String connectionQuality; // 'excellent', 'good', 'fair', 'poor'
  final DateTime createdAt;
  final DateTime updatedAt;

  Call({
    required this.id,
    this.caller,
    this.callee,
    required this.type,
    required this.status,
    this.startedAt,
    this.endedAt,
    this.duration = 0,
    this.isScreenSharing = false,
    this.screenShareStartedAt,
    this.connectionQuality = 'good',
    required this.createdAt,
    required this.updatedAt,
  });

  factory Call.fromJson(Map<String, dynamic> json) => _$CallFromJson(json);
  Map<String, dynamic> toJson() => _$CallToJson(this);

  String getDurationString() {
    if (duration == 0) return 'Not connected';

    final hours = duration ~/ 3600;
    final minutes = (duration % 3600) ~/ 60;
    final seconds = duration % 60;

    if (hours > 0) {
      return '${hours}h ${minutes}m ${seconds}s';
    } else if (minutes > 0) {
      return '${minutes}m ${seconds}s';
    } else {
      return '${seconds}s';
    }
  }

  String getStatusDisplay() {
    switch (status) {
      case 'initiated':
        return 'Calling...';
      case 'ringing':
        return 'Ringing...';
      case 'accepted':
        return 'Connected';
      case 'rejected':
        return 'Rejected';
      case 'ended':
        return 'Ended';
      case 'missed':
        return 'Missed';
      case 'busy':
        return 'Busy';
      default:
        return status;
    }
  }

  bool get isActive =>
      status == 'initiated' || status == 'ringing' || status == 'accepted';
  bool get isVoiceCall => type == 'voice';
  bool get isVideoCall => type == 'video';
}

File: elearningit/backend/models/Call.js
// models/Call.js
const mongoose = require('mongoose');

const callSchema = new mongoose.Schema({
  caller: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  callee: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  type: {
    type: String,
    enum: ['voice', 'video'],
    required: true
  },
  status: {
    type: String,
    enum: ['initiated', 'ringing', 'accepted', 'rejected', 'ended', 'missed', 'busy'],
    default: 'initiated'
  },
  startedAt: {
    type: Date
  },
  endedAt: {
    type: Date
  },
  duration: {
    type: Number, // in seconds
    default: 0
  },
  isScreenSharing: {
    type: Boolean,
    default: false
  },
  screenShareStartedAt: {
    type: Date
  },
  connectionQuality: {
    type: String,
    enum: ['excellent', 'good', 'fair', 'poor'],
    default: 'good'
  }
}, {
  timestamps: true
});

// Calculate duration when call ends
callSchema.pre('save', function(next) {
  if (this.endedAt && this.startedAt) {
    this.duration = Math.floor((this.endedAt - this.startedAt) / 1000);
  }
  next();
});

// Get call history for a user
callSchema.statics.getHistory = async function(userId, limit = 50) {
  return this.find({
    $or: [
      { caller: userId },
      { callee: userId }
    ]
  })
    .populate('caller', 'firstName lastName username profilePicture')
    .populate('callee', 'firstName lastName username profilePicture')
    .sort({ createdAt: -1 })
    .limit(limit);
};

// Get active calls for a user
callSchema.statics.getActiveCalls = async function(userId) {
  return this.find({
    $or: [
      { caller: userId },
      { callee: userId }
    ],
    status: { $in: ['initiated', 'ringing', 'accepted'] }
  })
    .populate('caller', 'firstName lastName username profilePicture')
    .populate('callee', 'firstName lastName username profilePicture');
};

module.exports = mongoose.model('Call', callSchema);
